### PR TITLE
feat(pkg/std): ensure files are sorted in a `MemPackage`

### PR DESCRIPTION
**Pull request by waymobetta**
This PR ensures in `MemPackage.Validate` that the files that it is being passed are correctly sorted. This is called in many places, among which the VM keeper.

**Context:**
It was discovered that during the add package process, gno files get processed in an unpredictable order. 

This relates to the below related issue (#1482) regarding multiple `init` statements being allowed since these `init` statements would potentially be processed in unknown order.

Related: https://github.com/gnolang/gno/issues/1482

<details><summary>Contributors' checklist...</summary>

- [ ] Added new tests, or not needed, or not feasible
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ ] Updated the official documentation or not needed
- [ ] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [x] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

### CODE REVIEW

**Review by deelawn — COMMENTED** [review]
I think perhaps the only change necessary is to add something like this:
```
sort.Slice(
    msg.Package.Files,
    func(i, j int) bool {
        return msg.Package.Files[i].Name < msg.Package.Files[j].Name
    },
)
```

That should sort the package files in place as long as the original order doesn't need to be preserved, but I don't think it does.

**Comment by waymobetta** [comment]
> I think perhaps the only change necessary is to add something like this:
> 
> ```
> sort.Slice(
>     msg.Package.Files,
>     func(i, j int) bool {
>         return msg.Package.Files[i].Name < msg.Package.Files[j].Name
>     },
> )
> ```
> 
> That should sort the package files in place as long as the original order doesn't need to be preserved, but I don't think it does.

I was thinking copy and replace versus altering in place for safety but this is a significantly more elegant solution.

File: gno.land/pkg/sdk/vm/keeper.go
Comment (deelawn) [inline gno.land/pkg/sdk/vm/keeper.go:180]: Is this link necessary?  I think it should be straightforward enough for someone to find the sort package documentation if they aren't familiar.

**Review by deelawn — APPROVED** [review]
Looks good to me.  One more thing to maybe consider -- this seems to be the correct way to sort strings, but does is the sorting exactly the same as what happens when go parses files?  Probably.  But how do we know for sure?

File: gno.land/pkg/sdk/vm/keeper.go
Comment (waymobetta) [reply gno.land/pkg/sdk/vm/keeper.go:180]: From what I can find, Go parses files _lexically_ and returns the resulting AST. As for building, perhaps more research is needed to look into the action graph for Go, but from my research this dependency tree's order is alphabetical, as shown [here (build.ReadDir)](https://cs.opensource.google/go/go/+/refs/tags/go1.21.6:src/go/build/build.go;l=106) and [here (os.ReadDir)](https://cs.opensource.google/go/go/+/refs/tags/go1.16beta1:src/os/dir.go;l=111).

[This](https://cs.opensource.google/go/go/+/refs/tags/go1.21.6:src/sort/sort.go;l=21-23) bit about the comparison operation is interesting and could prove to be an issue, so we may need to replace with a more stable sort variant instead to ensure order remains consistent in all cases. Having a test or two written may address the concern, though I wasn't sure how best to go about that.

> 	// If both Less(i, j) and Less(j, i) are false,
	// then the elements at index i and j are considered equal.
        // Sort may place equal elements in any order in the final result

**Review by thehowl — COMMENTED** [review]
I think to address your comment about ordering and doing a small performance improvement, too:

- Move the sorting into `std.MemPackage.Validate` (this will eventually be moved to `sdk/vm`, [per this comment](https://github.com/gnolang/gno/pull/1393) discussed with Jae)
- After sorting, change the current check for duplicates in `Validate()` (the one using a map) to a simpler one:

```go
// NOTE: Before this, add assertion that MemPkg contains at least 1 file

prev := mempkg.Files[0].Name
for _, file := range mempkg.Files[1:] {
	if prev == file.Name {
		// return error
	}
	prev = file.Name
}
```

(The following works because the slice is sorted)

**Review by moul — APPROVED** [review]

**Comment by waymobetta** [comment]
> I think to address your comment about ordering and doing a small performance improvement, too:
> 
> * Move the sorting into `std.MemPackage.Validate` (this will eventually be moved to `sdk/vm`, [per this comment](https://github.com/gnolang/gno/pull/1393) discussed with Jae)
> * After sorting, change the current check for duplicates in `Validate()` (the one using a map) to a simpler one:
> 
> ```go
> // XXX: add assertion that MemPkg contains at least 1 file
> prev := mempkg.Files[0].Name
> for _, file := range mempkg.Files[1:] {
> 	if prev == file.Name {
> 		// return error
> 	}
> 	prev = file.Name
> }
> ```
> 
> (The following works because the slice is sorted)

@thehowl Change to your suggestion or merge as is based on previous approvals?

**Comment by thehowl** [comment]
@waymobetta change it to mine, we get to kill two birds with one stone

**Review by thehowl — APPROVED** [review]
I fixed up this PR to move it to validate that the files are sorted instead. This seems more coherent with the line of thought of `Validate()` in general (ie. being readonly); rejecting unsorted MemPackage seems reasonable as there is currently no way to craft a MemPackage through official tools (like gnokey) where the filenames are not sorted.

I'll wait for another review, as my changes significantly change Jon's original work.

File: gno.land/pkg/sdk/vm/keeper.go
Comment (deelawn) [inline gno.land/pkg/sdk/vm/keeper.go:178]:
```suggestion
	pkgNames := make([]string, 0, len(msg.Package.Files))
```

**Pull request merged**
